modules/dnn/src/darknet/darknet_io.cpp

```diff
@@ -574,17 +672,40 @@ namespace cv {
                             batch_normalize);
 
                         current_channels = filters;
+                        current_width = (current_width - kernel_size + 2 * padding) / stride + 1;
+                        current_height = (current_height - kernel_size + 2 * padding) / stride + 1;
+                    }
+                    else if (layer_type == "connected")
+                    {
+                        int output = getParam<int>(layer_params, "output", 1);
+                        bool batch_normalize = getParam<int>(layer_params, "batch_normalize", 0) == 1;
+
+                        CV_Assert(output > 0);
+
+                        setParams.setFullyConnected(output, batch_normalize);
+
+                        if(layers_counter && current_width)
+                            net->out_channels_vec[layers_counter-1] = current_channels * current_width * current_height;
+
+                        current_channels = output;
+                        current_width = 0;
+                        current_height = 0;
```

Comment:
What if network will have two subsequent `connected` layers? A product will be zero.
Maybe it's better to have `MatShape` instead of `current_channels`, `current_width` and  `current_height` with last layer's output shape. This way it can be variable size.

Reply:
Sorry, couldn't understand this.

> A product will be zero.

https://github.com/opencv/opencv/blob/c15b33ef80e0fc811b9c9d285d8db8211659da25/modules/dnn/src/darknet/darknet_io.cpp#L687-L690

Using this if statement we are checking the condition that the current layer should not be the first layer of model `layer_counter != 0` and last layer was not connected layer `current_width != 0`. Only then we are doing product and updating that product in previous layer's (`layers_counter-1`) `out_channel_vec`.
`net->out_channels_vec[layers_counter-1] = current_channels * current_width * current_height;`

If current connected is not the first connected layer, it will simply update `out_channel_vec` by `current_channels` (i.e. `output` of current connected layer)
https://github.com/opencv/opencv/blob/c15b33ef80e0fc811b9c9d285d8db8211659da25/modules/dnn/src/darknet/darknet_io.cpp#L690
https://github.com/opencv/opencv/blob/c15b33ef80e0fc811b9c9d285d8db8211659da25/modules/dnn/src/darknet/darknet_io.cpp#L814

For test, [connected.cfg](https://github.com/opencv/opencv_extra/blob/c3db6200e6c87e42f192f45a61f604ae6562b918/testdata/dnn/darknet/connected.cfg) has two connected layers.

Reply:
Got it. I just wanted to solve two problems at the same time - remove this condition because it's too specific and remove all the calculations of width/height from onnx_importer but replace them to layers' `getMemoryShapes`. So if you could now just introduce MatShape and use it, we will easier integrate `getMemoryShapes` in separate PR.

Reply:
So for now it's OK to compute width and height. But If you can replace it by MatShape - it'd be simpler in the future.

Reply:
sure.